nodeJS/express/express_forms_validation.md

```diff
@@ -0,0 +1,565 @@
+
+usersRouter.get("/create", usersController.usersCreateGet);
+usersRouter.post("/create", usersController.usersCreatePost);
+
+module.exports = usersRouter;
+```
+
+Next we'll create two views:
+
+- `index.ejs` will list all the users we've created.
+- `users.ejs` will display our main form.
+
+```ejs
+<!-- views/index.ejs -->
+<!DOCTYPE html>
+<html>
+<body>
+<h1><%= title %></h1>
+<ul>
+  <% if (locals.users) {%>
+    <% users.forEach(user => { %>
+      <li>
+        ID: <%= user.id %>, Name: <%= user.firstName %> <%= user.lastName %>
+      </li>
+    <% }); %>
+  <% } %>
+</ul>
+<a href="/users/create">Create a user</a>
+</body>
+</html>
+```
+
+```ejs
+<!-- views/users.ejs -->
+<!DOCTYPE html>
+<html>
+<head>
+  <title><%= title %></title>
+</head>
+<body>
+<h1><%= title %></h1>
+<form action="/users/create" method="POST">
+  <label for="firstName">First Name: </label>
+  <input type="text" name="firstName" required>
+  <label for="lastName">Last Name: </label>
+  <input type="text" name="lastName" required>
+  <button type="submit">Submit</button>
+</form>
+<a href="/">Back to home</a>
+</body>
+</html>
+```
+
+The logic for this router will go inside of our controller:
+
+```javascript
+// controllers/usersController.js
+// User is a factory function which acts as a model for our simulated database.
+// We will explore this concept more as we later dive deeper into databases.
+const User = (firstName, lastName) => ({ firstName, lastName });
+const asyncHandler = require("express-async-handler");
+const usersStorage = require("../storages/usersStorage");
+
+exports.usersCreateGet = asyncHandler(async (req, res) => {
+  res.render("users", { title: "User List", users: usersStorage.getUsers() });
+});
+
+exports.usersCreatePost = asyncHandler(async (req, res) => {
+  const { firstName, lastName } = req.body;
+  usersStorage.addUser(User(firstName, lastName));
+  res.redirect("/");
+});
+```
+
+And we'll use a storage class to hold the users we create. In real-world scenarios, you would almost certainly be using a database for this, which you'll explore further in upcoming lessons.
+
+```javascript
+// storages/usersStorage.js
+// This class lets us simulate interacting with a database.
+class UsersStorage {
+    constructor() {
+        this.storage = {};
+        this.id = 0;
+    }
+
+    addUser({firstName, lastName}) {
+        const id = this.id;
+        this.storage[id] = { id, firstName, lastName };
+        this.id++;
+    }
+
+    getUsers() {
+        return Object.values(this.storage);
+    }
+
+    getUser(id) {
+        return this.storage[id];
+    }
+
+    updateUser(id, {firstName, lastName}) {
+        this.storage[id] = { id, firstName, lastName };
+    }
+
+    deleteUser(id) {
+        delete this.storage[id];
+    }
+}
+// Rather than exporting the class, we can export an instance of the class by instantiating it.
+// This ensures only one instance of this class can exist, also known as the "singleton" pattern.
```

Comment:
Super cool addition. Content like this really makes our lessons stand out. Good job!